develop/tutorials/articles/310-troubleshooting/03-resolving-cnfe-and-cdnfe-in-osgi-bundles.markdown

```diff
@@ -4,74 +4,116 @@ Understanding a `ClassNotFoundException` or `NoClassDefFoundError` in non-OSGi
-`NoClassDefFoundError` occurs.
-
-In the consuming bundle, make sure to import the correct package. Check the
-package name. If the package import is correct but you still get the exception
-or error, the class might no longer exist in the package.
-
-## Case 2: Class no longer exists in the package [](id=case-2-class-no-longer-exists-in-the-package)
-
-In OSGi runtime environments, bundles can change and come and go. If you
-reference or look up another bundle's class and that class has been removed, a
-`NoClassDefFoundError` or `ClassNotFoundException` occurs.
-[Semantic Versioning](http://semver.org) guards against this scenario. It
-dictates that removing a class from an exported package constitutes a new major
-version for that package. Neglecting to increment the package's major version
-breaks dependent bundles. 
-
-For example, a bundle that consumes class `com.foo.Bar` specifies package import
-`com.foo;version=[1.0.0, 2.0.0)`--the bundle uses `com.foo` package versions
-from `1.0.0` up to but not including `2.0.0`. The first part of the version
-number (the `1` in `1.0.0`) represents the *major* version. The consuming bundle
-doesn't expect any *major* breaking changes, such as a class being removed.
-Removing class `com.foo.Bar` from package `com.foo` constitutes a major breaking
-change. Removing the class without incrementing the major version number breaks
-consuming bundles. 
-
-The example consuming bundle doesn't expect any major breaking changes, such as
-a class being removed. If the authoring developer removes class `Bar` from
-`com.foo` without incrementing the package to a new major version (e.g.,
-`2.0.0`), a `ClassNotFoundException` or `NoClassDefFoundError` occurs when other
-bundles look up or reference that class.
-
-Here are a couple options to consider since the class no longer exists in the package:
-
--   Adapt to the new API. To learn how to adapt, read the package's/module's
-    Javadoc, release notes, and or formal documentation. You can ask the author
-    or investigate forums. 
-
--   Revert to the module version you used previously. Deployed module versions
-    reside in `[Liferay_Home]/osgi/`. See
-    [Backing up Liferay Installations](https://dev.liferay.com/discover/deployment/-/knowledge_base/7-0/backing-up-a-liferay-installation#backing-up-liferays-file-system)
-    for details. 
-
-Do what you think is best to get your module working properly. 
-
-Now you know how to resolve common situations involving `ClassNotFoundException`
-or `NoClassDefFoundError`. For additional information on `NoClassDefFoundError`
-see OSGi Enroute's article
-[What is NoClassDefFoundError?](http://enroute.osgi.org/faq/class-not-found-exception.html).
+    classpath or using an invalid name to look up a class. 
+-   `NoClassDefFoundError`: occurs when a compiled class references another 
+    class that isn't on the runtime classpath. 
+
+In OSGi environments, however, things are more complex. When trying to determine 
```

Comment:
I purposely got rid of the "more complex" language last time in my wordsmith. Why are we re-adding it?